appserver/payara-appserver-modules/microprofile/fault-tolerance/src/main/java/fish/payara/microprofile/faulttolerance/policy/FaultTolerancePolicy.java

```diff
@@ -0,0 +1,609 @@
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.microprofile.faulttolerance.policy;
+
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.interceptor.InvocationContext;
+
+import org.eclipse.microprofile.faulttolerance.exceptions.BulkheadException;
+import org.eclipse.microprofile.faulttolerance.exceptions.CircuitBreakerOpenException;
+import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException;
+import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceException;
+import org.eclipse.microprofile.faulttolerance.exceptions.TimeoutException;
+
+import fish.payara.microprofile.faulttolerance.FaultToleranceConfig;
+import fish.payara.microprofile.faulttolerance.FaultToleranceService;
+import fish.payara.microprofile.faulttolerance.FaultToleranceMetrics;
+import fish.payara.microprofile.faulttolerance.state.BulkheadSemaphore;
+import fish.payara.microprofile.faulttolerance.state.CircuitBreakerState;
+
+/**
+ * The {@link FaultTolerancePolicy} describes the effective aggregated policies to use for a particular {@link Method}
+ * when adding fault tolerant behaviour to it.
+ * 
+ * The policies are extracted from FT annotations and the {@link FaultToleranceConfig}.
+ * 
+ * In contrast to the plain annotations the policies do consider configuration overrides and include validation of the
+ * effective values.
+ * 
+ * The policy class also reduces the need to analyse FT annotations for each invocation and works as a consistent source
+ * of truth throughout the processing of FT behaviour that is convenient to pass around as a single immutable value.
+ *
+ * @author Jan Bernitt
+ */
+public final class FaultTolerancePolicy implements Serializable {
+
+    static final Logger logger = Logger.getLogger(FaultTolerancePolicy.class.getName());
+
+    private static final long TTL = 60 * 1000;
+
+    /**
+     * A simple cache with a fix {@link #TTL} with a policy for each target method.
+     */
+    private static final ConcurrentHashMap<Class<?>, ConcurrentHashMap<Method, FaultTolerancePolicy>> POLICY_BY_METHOD 
+        = new ConcurrentHashMap<>();
```

Comment:
Originally this was intentionally done this way to avoid object creation on lookup. I was hoping to not have to create garbage for each invocation. Later this turned out to be very difficult so there will be 2-3 garbage objects per invocation. We could change this and make it a bit more garbage and it certainly would increase readability. On the other hand there are just 2 methods using this so the simplification isn't that big and not creating an additional object might still be worth it. WDYT?

Reply:
My gut feeling says don't optimise at this level until it's really proven to be a bottleneck (e.g. this being in a tight loop with thousands of lookups). In CDI these temp objects might be dwarfed by all the other things going on.

Also, in select cases the JVM will allocate objects on the stack (if escape analysis proves they can't escape their scope), making object creation for temp object very cheap).

Reply:
It was not really applied optimizing. I saw two options: use a key class or do a nested map. I knew there was only two usages of the structure and nested maps had the benefit of avoiding garbage objects so I went for that option. Making this decision a big thing is maybe also wrong focus. Only reason I did not change it was that it does not make much sense to put more work into something that makes so little difference as its scope is and will be tiny.

Reply:
I was more looking at this from code readability point of view, as operations on nested maps were always bit hard to read (although it's now better with `computeIfAbsent`).

A theoretical performance argument for the key class would be, that it will reduce lookup time, by only utilizing single map rather than two of them.

A compromise solution would be to encapsulate the map into separate class only exposing the operations that are needed.

Reply:
I was thinking about readability was well. So ultimately the question now is: do we think improved readability is worth the effort of changing this, then I think we should do it. I hesitated since there are literally 2 lines of code affected and using a key class would mean actually more code line wise so it felt less clear and didn't really call for action so I left it alone.